logserver/src/main/java/alluxio/logserver/AlluxioLogServerProcess.java

```diff
@@ -0,0 +1,228 @@
+        break;
+      }
+    }
+  }
+
+  /**
+   * Called from another thread (actually in the shutdown hook of {@link AlluxioLogServer}
+   * to stop the main thread of {@link AlluxioLogServerProcess}.
+   */
+  private void stopServing() {
+    mStopped = true;
+    if (mServerSocket != null) {
+      try {
+        mServerSocket.close();
+      } catch (IOException e) {
+        LOG.warn("Exception in closing server socket.", e);
+      }
+    }
+
+    mThreadPool.shutdown();
+    long timeoutMS = STOP_TIMEOUT_IN_MS;
+    long now = System.currentTimeMillis();
+    while (timeoutMS >= 0) {
+      try {
+        mThreadPool.awaitTermination(timeoutMS, TimeUnit.MILLISECONDS);
+        break;
+      } catch (InterruptedException e) {
+        long newnow = System.currentTimeMillis();
+        timeoutMS -= (newnow - now);
+        now = newnow;
+      }
+    }
+  }
+
+  /**
+   * Configure a {@link Hierarchy} instance used to retrive logger by name and maintain the logger
+   * hierarchy. An instance of this class will be passed to a {@link AlluxioLog4jSocketNode} so
+   * that the {@link AlluxioLog4jSocketNode} instance can retrieve the logger to log incoming
+   * {@link org.apache.log4j.spi.LoggingEvent}s.
+   *
+   * @param inetAddress inet address of the client
+   * @param logAppenderName name of the appender to use for this client
+   * @return A {@link Hierarchy} instance to pass to {@link AlluxioLog4jSocketNode}
+   * @throws IOException if fails to create an {@link FileInputStream} to read log4j.properties
+   * @throws URISyntaxException if fails to derive a valid URI from the value of property named
+   *         "log4j.configuration"
+   */
+  protected LoggerRepository configureHierarchy(InetAddress inetAddress, String logAppenderName)
+      throws IOException {
+    Hierarchy clientHierarchy;
+    String inetAddressStr = inetAddress.getHostAddress();
+    Properties properties = new Properties();
+    try {
+      final File configFile = new File(new URI(System.getProperty("log4j.configuration")));
+      try (FileInputStream inputStream = new FileInputStream(configFile)) {
+        properties.load(inputStream);
+      }
+    } catch (URISyntaxException e) {
+      // Alluxio log server cannot derive a valid path to log4j.properties. Since this
+      // properties file is global, we should throw an exception.
+      throw new RuntimeException(e);
+    }
+    Level level = Level.INFO;
+    clientHierarchy = new Hierarchy(new RootLogger(level));
+    // Startup script should guarantee that mBaseLogsDir already exists.
+    String logDirectoryPath = mBaseLogsDir + "/" + logAppenderName.toLowerCase();
+    File logDirectory = new File(logDirectoryPath);
+    LOG.info(logDirectoryPath);
+    if (!logDirectory.exists()) {
+      logDirectory.mkdir();
+    }
+    String logFilePath = logDirectoryPath + "/" + inetAddressStr + ".log";
+    properties.setProperty("log4j.rootLogger", level.toString() + ",LOGSERVER_CLIENT_LOGGER");
```

Comment:
Define constants for important string constants like `LOGSERVER_CLIENT_LOGGER`, and comment on the requirement that they need to match the `log4j.properties` file

Reply:
Done.